src/Components/GEOSldas_GridComp/GEOSlandassim_GridComp/clsm_ensupd_read_obs.F90

```diff
@@ -1548,7 +1548,461 @@ subroutine read_obs_sm_ASCAT(                                  &
+    real*8,    dimension(:),     allocatable    :: tmp1_jtime
+    real*8,    dimension(:,:),   allocatable    :: tmp_data
+
+    real,      dimension(:),     pointer        :: tmp_obs,  tmp_lat,  tmp_lon
+    real*8,    dimension(:),     pointer        :: tmp_jtime
+
+    integer,   dimension(:),     pointer        :: tmp_tile_num
+
+    integer,   dimension(N_catd)                :: N_obs_in_tile    
+
+    character(len=*),  parameter :: Iam = 'read_obs_sm_ASCAT_EUMET'
+    character(len=400)           :: err_msg
+
+    ! -------------------------------------------------------------------
+    
+    nullify( tmp_obs, tmp_lat, tmp_lon, tmp_tile_num, tmp_jtime )    
+    
+    ! ---------------
+    
+    ! initialize
+    
+    found_obs = .false.
+    
+    ! find files that are within half-open interval 
+    !   [date_time-dtstep_assim/2,date_time+dtstep_assim/2)
+
+    date_time_low = date_time    
+    call augment_date_time( -(dtstep_assim/2), date_time_low)
+    date_time_up = date_time    
+    call augment_date_time(  (dtstep_assim/2), date_time_up)
+    
+    ! calculate "extra" date_time_low to catch files w/ time swath start stamps before window 
+    !   but containing relevant obs
+    
+    date_time_low_fname = date_time_low
+    call augment_date_time( -dt_ASCAT_obsfile, date_time_low_fname)
+    
+    ! ----------------------------------------------------------------
+    !
+    ! read file with list of ASCAT file names for first day
+    
+    fname_of_fname_list = 'dummy'  ! make sure it is properly defined in obs_param nml
+    
+    call read_obs_SMAP_fnames( date_time_low_fname, this_obs_param,            &
+         fname_of_fname_list, N_fnames_max,                                    &
+         N_fnames, fname_list(1:N_fnames_max) )
+    
+    ! if needed, read file with list of ASCAT file names for second day and add
+    !   file names into "fname_list"
+    
+    if (date_time_low_fname%day /= date_time_up%day) then
+       
+       call read_obs_SMAP_fnames( date_time_up, this_obs_param,                &
+            fname_of_fname_list, N_fnames_max,                                 &
+            N_fnames_tmp, fname_list((N_fnames+1):(N_fnames+N_fnames_max)) )
+       
+       N_fnames = N_fnames + N_fnames_tmp
+       
+    end if
+    
+    tmpfnames = fname_list(1:N_fnames)
+    
+    ! ----------------------------------------------------------------
+    !
+    ! find files that have obs within assimilation window
+    
+    N_tmp = 0   
+    
+    do kk = 1,N_fnames
+
+       tmpfname = fname_list(kk)
+
+       ! Are we in the required assimilation window?
+       !
+       ! e.g. Y2019/M07/M01-ASCA-ASCSMO02-NA-5.0-20190702075700.000000000Z-20190702084627-1350204.bfr
+       !
+       !      12345678901234567890123456789012345678901234567890123456789012345678901234567890 
+       !               1         2         3         4         5         6         7
+
+       str_date_time = tmpfname(40:53)
```

Comment:
I don't know where the list files are and couldn't check the string processing here.  The sample "fname" probably needs to be updated.  Are the list files in day directories?  Or can/should we have the date info in the list file name?  This probably requires modifying the SMAP-specific subroutine that handles the list files.  (We may need this for the MODIS snow cover obs as well.)  Would it help to have the BUFR obs files in day directories?

Reply:
This was a hack, fixed by:
- read_obs_SMAP_fnames -> [read_obs_fnames](https://github.com/GEOS-ESM/GEOSldas/blob/3f95ee02780cf30259b9a2ea85eceb000822a7e7/src/Components/GEOSldas_GridComp/GEOSlandassim_GridComp/clsm_ensupd_read_obs.F90#L7149)
- optional argument [obs_dir_hier](https://github.com/GEOS-ESM/GEOSldas/blob/3f95ee02780cf30259b9a2ea85eceb000822a7e7/src/Components/GEOSldas_GridComp/GEOSlandassim_GridComp/clsm_ensupd_read_obs.F90#L7174C1-L7174C66) in read_obs_fnames to write YYYY/MM/filename
```fortran
          if (present(obs_dir_hier) .and. obs_dir_hier == 1) then
            
            ! preface file names with "Yyyyy/Mmm"
            
            fname_list(ii) = YYYYMMdir // trim(tmpstr80)


          elseif (present(obs_dir_hier) .and. obs_dir_hier /= 1) then
            
            err_msg = 'Unrecognized obs_dir_hier #'
            call ldas_abort(LDAS_GENERIC_ERROR, Iam, err_msg)


          else
            ! preface file names with "Yyyyy/Mmm/Ddd" Default for SMAP obs
            
            fname_list(ii) = YYYYMMDDdir // trim(tmpstr80)
          
          end if
```
- update date from string [parse]( https://github.com/GEOS-ESM/GEOSldas/blob/3f95ee02780cf30259b9a2ea85eceb000822a7e7/src/Components/GEOSldas_GridComp/GEOSlandassim_GridComp/clsm_ensupd_read_obs.F90#L1726C8-L1726C39) to reflect this change